### ISSUE TITLE
 Alternator's response to oversized request may not reach the client because of an RST

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#12166]
Alternator places a limit (using Seastar's HTTP server feature of `content_length_limit`) on the request size. 

We have two tests,
```
test_manual_requests.py::test_too_large_request_content_length
test_manual_requests.py::test_too_large_request_chunked
```
which check what happens when the request's length exceeds the limit. We expect the client to be able to read the HTTP error response `413 Payload Too Large`. This **usually** works correctly and the first test usually passes (the second fails because of an unrelated issue). But because of a Seastar bug https://github.com/scylladb/seastar/issues/1325 sometimes the client receives a RST packet before it can read the response, and these tests fail. The failure is highly dependent on timing - some people seem to be able to reproduce this failure fairly frequently, while others can run the test thousands of times without failure.

I am opening this ScyllaDB issue for waiting for the Seastar issue, https://github.com/scylladb/seastar/issues/1325, to be fixed.

**Comment by nyh** [comment]
@avikivity of course, I preparing now a patch that will XFAIL the test and mark issue #12143. This issue will stay open until the test is fixed.

**Comment by DoronArazii** [comment]
@nyh can we close it?

**Comment by nyh** [comment]
Not really, the bug still exists. It's not very serious, and not common, but it was never fixed.

**Comment by mykaul** [comment]
> @mykaul instead of pushing it from version to version, lets put it in the backlog and wait for commitment. otherwise we will visit here every 3 months.

I use the '3 strikes' method. If it missed 3 releases, I'll put it on the backlog / close.

**Comment by nyh** [comment]
Note that issue https://github.com/scylladb/scylladb/issues/8196 suggests we should stop using Seastar's content_length_limit feature - which is too limited (e.g., currently it doesn't work for chunked encoding), and instead have Scylla implement this feature. If we do this, we still need to consider how to handle this situation in Scylla - but the fix will be in Scylla, not in Seastar. The various comments made in https://github.com/scylladb/seastar/issues/1325 will be relevant in an in-Scylla implementation as well:
1. To avoid RST, when the request read has become too large, we should send an error response, half-close the connection, and continue to read the request.
2. However, after reading "enough" more garbage, we should consider closing the connection completely. We don't want the user to be able to send us 1 GB or data if we limited the request size to 16 MB. Even if it means the application will lose the error message in that unlikely case (I'm hoping that the test cases, not involving 1 GB of data,  will not fail).

**Comment by nyh** [comment]
Pull request https://github.com/scylladb/scylladb/pull/23434 changed the way that Scylla limits a request length - we no longer use Seastar's content_length_limit feature, but rather do the limitation inside Scylla - while reading the request, if we surpassed 16 MB of body, we immediately return an error.

This is still exposed to https://github.com/scylladb/seastar/issues/1325 because as explained there, when Seastar sees a connection for which we have not read the entire request body, it closes the connection - and can result in the RST problem.

So we still need https://github.com/scylladb/seastar/issues/1325 to be fixed in Seastar for this issue to be fixed.
